lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_eager_fallback.cpp

```diff
@@ -0,0 +1,70 @@
+#include "lazy_tensor_core/csrc/ts_backend/aten_eager_fallback.h"
+
+#include <unordered_map>
+
+#include "lazy_tensor_core/csrc/function_call_tracker.h"
+#include "lazy_tensor_core/csrc/ts_backend/EagerFallback.h"
+#include "lazy_tensor_core/csrc/ts_backend/ts_computation_client.h"
+#include "lazy_tensors/computation_client/debug_macros.h"
+#include "lazy_tensors/computation_client/ltc_logging.h"
+#include "lazy_tensors/computation_client/metrics.h"
+
+namespace torch_lazy_tensors {
+namespace {
+
+c10::Backend TSHardwareBackend() {
+  c10::DeviceType device_type =
+      lazy_tensors::compiler::TSComputationClient::HardwareDeviceType();
+  switch (device_type) {
+    case c10::DeviceType::CPU: {
+      return c10::Backend::CPU;
+    }
+    case c10::DeviceType::CUDA: {
+      return c10::Backend::CUDA;
+    }
+    default: {
+      LTC_LOG(FATAL) << "Backend not found for device type: " << device_type;
+    }
+  }
+}
+
+}  // namespace
+
+static std::unordered_map<std::string, ::lazy_tensors::metrics::Counter*>
+    _eager_fallback_counters;
+
+void ltc_eager_fallback(const c10::OperatorHandle& op,
+                        torch::jit::Stack* stack) {
+  LTC_FN_TRACK(3);
+  const auto name = c10::toString(op.operator_name());
+
+  // Manually applying the LTC_COUNTER macro.
+  // We need to do it ourselves and explicitly keep a mapping of counters
+  // because this boxed fallback kernel is used by multiple operators,
+  // and the macro stamps out a static Counter object with a fixed name
+  // at the code location that it was called.
+  if (_eager_fallback_counters.find(name) == _eager_fallback_counters.end()) {
+    _eager_fallback_counters[name] = new ::lazy_tensors::metrics::Counter(name);
+  }
+  _eager_fallback_counters[name]->AddValue(1);
+
+  auto& args = op.schema().arguments();
+  auto arguments = torch::jit::last(stack, args.size());
+
+  // Log each tensor argument.
+  for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+    const auto& ivalue = arguments[idx];
+    if (ivalue.isTensor()) {
+      LTC_VLOG(3) << ivalue.toTensor().toString();
+    }
+  }
+
+  // Call the actual boxed CPU fallback.
+  eager_fallback(op, stack, TSHardwareBackend());
+}
+
+TORCH_LIBRARY_IMPL(_, XLA, m) {
```

Comment:
Everything in this file seems like something that should go into the xla plugin - not in the in-tree ts_backend integration.

I don't have a full understanding of how ts_backend is used. My understanding is that:

(a) LTC is a tool, and isn't in the business of registering kernels to the dispatcher itself (it doesn't have a DispatchKey, after all). Instead, vendors that utilize LTC will be the ones registering kernels to the dispatcher, under their particular dispatch key. Those kernels will call out to the LTC API.

(b) Similar, I thought that ts_backend integration functions the same way - it doesn't register anything to the dispatcher directly, and instead the vendors that integrate with it will be the ones registering to the dispatcher.

And kinda related - it seems incorrect that we're registering anything to the XLA dispatch key within the TS integration, right?

Reply:
Your understanding is right, but we have a TorchScript backend we bundle with LTC. It isn't enabled by default, you have to call `lazy_tensor_core._LAZYC._ltc_init_ts_backend()` to register it.

A vendor which doesn't want the TorchScript backend will register theirs and as long as they don't call `lazy_tensor_core._LAZYC._ltc_init_ts_backend()` there's no conflict, I've tested it. Such a vendor can simply ignore the ts_backend folder.

Reply:
The `_ltc_init_ts_backend` makes sense, thanks for pointing it out.

Even in that case, wouldn't XLA be the wrong key for Torchscript to register the fallback to though?

And - this is me trying to get a better understanding of the integration - is TorchScript really a "backend" in this case? I'm thinking of a backend as an actual piece of hardware that can run the computation - shouldn't that be the thing doing the kernel registrations?

My understanding of torchscript vendor plugins is pretty vague, but I remember learning that NV TensorRT is a vendor plugin that integrates with TorchScript.

So if I'm following correctly, wouldn't the chain go "LTC graph" ---> (using LTC-TS plugin) ---> "TorchScript graph" ---> (using TS-TensorRT integration) ---> "TensorRT graph" ---> custom TensorRT hardware lowering? And then the TensoRT plugin would be responsible for registering the fallback kernel (to whatever dispatch key the TensorRT kernels currently map to)?

Reply:
I don't think XLA would be the "wrong" key, but we're definitely hijacking it for something which isn't XLA, which could make it confusing. It's an ongoing discussion of what we should do about open key registration for various vendors, but hijacking it doesn't create problems - we'd never want multiple LTC backends to coexist, so you either register the XLA one or the TorchScript one but not both.

Your point about TorchScript (not) being a back-end is well taken, I think it equally applies to XLA. Calling it a back-end is defensible though, because you're coming from a Python front-end and a graph extraction middle-end (the lazy tensor core). XLA has its own way of controlling the physical device on which it's running, so if you wanted to register based on hardware target you'd have XLA:CPU, XLA:GPU and XLA:TPU devices (and create confusion with CPU and GPU as eager devices). Look at these back-ends as virtual devices, using real hardware which we don't have a good reason to expose.

Reply:
You're point about XLA and TS both being a graph-capture middle end makes sense. What seems like the main difference to me though is that XLA graph capture + lowering hardware like TPU is all bundled together, so users know that e.g. `torch.device('xla:1')` maps to some xla-aware hardware.

Versus with the TS integration, it sounds like we might have multiple vendors interested in integrating with TorchScript? And IIUC, the usage of the boxed fallback here + the `ts_native_functions.yaml` file imply that vendors who integrate with torchscript will be forced to use `torch.device('xla')` as their frontend API entry point. If different vendors want to be able to have their own devices in the frontend API, I think they'll need to be the ones creating the yaml + fallback?

If we only have one vendor interested in the TS integration for now and in the near future, and they're happy to use the XLA device, then this is probably fine. But otherwise, is it worth moving this into the vendor code?

Reply:
Not sure what you mean by "moving this into the vendor code". This fallback? We agree that it won't be in the vendor code.

Regarding the key / device, having an open registration mechanism for keys would be optimal. At the same time, I think we can live with a "multiplexed" lazy tensor device and key - it shouldn't be named XLA, of course. I don't think open registration discussion belongs to this PR, though.